cirq-core/cirq/sim/simulation_state.py

```diff
@@ -166,6 +166,37 @@ def create_merged_state(self) -> Self:
         """Creates a final merged state."""
         return self
 
+    def add_qubits(self: Self, qubits: Sequence['cirq.Qid']) -> Self:
+        """Add qubits to a new state space and take the kron product.
+        Note that only subclasses that support `kronecker_product`
+        will support this function. E.g Density Matrix and
+        State Vector simulators.
+
+        Args:
+            qubits: Sequence of qubits to be added.
+
+        Returns:
+            A new Simulation State with the new qubits added. Or
+            `self` if there are no qubits to add.
+        """
+        if qubits is None or not qubits:
+            return self
+        new_space = type(self)(qubits=qubits)
```

Comment:
@daxfohl In order to pass the typecheck test I need to pass in "state" which is a mandatory keyword. I assume I want to pass in a new state s.t it allocates space for the qubits I want to add.

Reply:
I think what you are looking for is 
`linalg.transformations.state_vector_kronecker_product` and `linalg.transformations.sub_state_vector`. but for `state_vector_kronecker_product` you need to know the state of the qubits. for clean qubits that $\ket{0 \cdots 0}$ but for borrowable qubits that can be anything. 

I think the starting state of borrowable qubit should be an input to the simulator. for example use this state as initial state for borrowable qubits. or we get a seed that we use to generate random initial states.

Reply:
@NoureldinYosri that would only be applicable to state vector simulators, but this is the base class so the state type could be anything.

@senecameeks yeah I vaguely recall now that being a challenge. I don't think there is a way currently to make a SimulationState abstractly without specifying the underlying QuantumState. At least not in a typesafe fashion. I guess options are

1. Create some abstract static factory method on SimulationState to create a zeroed out QuantumState (though then all subclasses would need to implement it so it would break backwards compatibility with third party simulators.) 
2. Make state optional in SimulationState constructor and default to zero. This is essentially the same thing and would also break backwards compatibility.
3. Pass in the QuantumState here as recommend by @NoureldinYosri. Assuming the ancilla qubits could be dirty and users may want to investigate the effects of dirty ancilla, this really is the only correct solution. But it'd likely require a completely different approach.
4. Just ignore the type check. Not great but it works fine for the standard use cases with clean ancillas and the normal simulators that have default zero-state constructors.

Per 3, maybe it uncovers a need do some deep refactoring where qubits are stateful objects in their own right rather than just indexes. Or pass the full state around everywhere. It's a big change, but maybe it's the only way to set up a proper foundation for these kinds of things.

Reply:
I say for now we return `NotImplemented` from the `add_qubits` method on the base class and override the method in two specific derived classes - state vector simulation state and density matrix simulation state. This is backwards compatible since any previously written user defined class just needs to implement a new method to add support for gates decomposing with ancillas and any previously written code continues to work. This is also type-safe. 


I've also added a condition to check that we only allocate space for ancilla qubits that are not already tracked by us. This is because we don't want to assume that an operation's decompose ḷalways allocates new unique ancillas. It can also use an existing system qubit as an ancilla qubit. See my explanation on https://github.com/quantumlib/Cirq/issues/6101#issuecomment-1577688779 for more details. 


@senecameeks Here is a commit that makes the suggested changes on top of your current PR. Can you please incorporate these changes into your PR and add necessary coverage tests ? 

https://github.com/tanujkhattar/Cirq/commit/ccde6892b01929036369f4313a85e707fe0338ac